packages/kit/src/runtime/client/client.js

```diff
@@ -640,52 +683,74 @@ export function create_client({ target, base, trailing_slash }) {
 		// to act on the failures at this point)
 		[...errors, ...layouts, leaf].forEach((loader) => loader?.().catch(() => {}));
 
-		const nodes = [...layouts, leaf];
+		const loaders = [...layouts, leaf];
 
 		// To avoid waterfalls when someone awaits a parent, compute as much as possible here already
-		/** @type {boolean[]} */
-		const nodes_changed_since_last_render = [];
-		for (let i = 0; i < nodes.length; i++) {
-			if (!nodes[i]) {
-				nodes_changed_since_last_render.push(false);
-			} else {
-				const previous = current.branch[i];
-				const changed_since_last_render =
-					!previous ||
-					(changed.url && previous.uses.url) ||
-					changed.params.some((param) => previous.uses.params.has(param)) ||
-					Array.from(previous.uses.dependencies).some((dep) => invalidated.some((fn) => fn(dep))) ||
-					(previous.uses.parent && nodes_changed_since_last_render.includes(true));
-				nodes_changed_since_last_render.push(changed_since_last_render);
+
+		let server_parent_changed = false;
+		let shared_parent_changed = false;
+
+		/** @type {Array<{ server: boolean, shared: boolean } | null>} */
+		const nodes_changed_since_last_render = loaders.map((loader, i) => {
```

Comment:
you could move this computation inside `if (route.uses_server_data)` so that you don't have to do it in other cases

Reply:
there's no point reviewing this in its current state @benmccann, it's a WIP that will completely change

Reply:
(or rather, has completely changed since i last pushed. e.g. the change you describe is something i did on friday night. bear with me)